README.md
# SwiftyPickerPopover

A more convenient way to display a popover with a built-in picker, on iPhone/iPad of iOS9+.

[![Version](https://img.shields.io/cocoapods/v/SwiftyPickerPopover.svg?style=flat)](http://cocoadocs.org/docsets/SwiftyPickerPopover)
[![License](https://img.shields.io/cocoapods/l/SwiftyPickerPopover.svg?style=flat)](http://cocoadocs.org/docsets/SwiftyPickerPopover)
[![Platform](https://img.shields.io/cocoapods/p/SwiftyPickerPopover.svg?style=flat)](http://cocoadocs.org/docsets/SwiftyPickerPopover)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/hsylife/SwiftyPickerPopover)

## Features

- By simple code, you can display a popover that contains a built-in picker, on iPhone or iPad.
- Swift 5, iOS9+. UIPopoverController free.
- Callback

## Screenshots

<img src="README_resources/SwiftyPickerPopover_movie.gif" width="308">

## Basic

```swift
DatePickerPopover(title: "DatePicker")
           .setDoneButton(action: { _, selectedDate in print(selectedDate) })
        .appear(originView: sender, baseViewController: self)
```

## Required

- Swift 5, Xcode 11.
- iOS 9+
- CocoaPods 1.1.0.rc.2+ or Carthage 0.12.0+

## License

MIT

## Install

### CocoaPods

Specify it in your 'Podfile', after replacing ‘YourProjectTargetName’ with your own target name:

```ruby
platform :ios, '9.0'
use_frameworks!
target ‘YourProjectTargetName’ do
  pod 'SwiftyPickerPopover'
end
```

Run 'pod install'.

### Carthage

- Add it to your Cartfile:

```
github "hsylife/SwiftyPickerPopover"
```

- Run `carthage update --platform iOS`
- Add 'SwiftyPickerPopover.framework' to 'Linked Frameworks and Library' on your project.
- Add `/usr/local/bin/carthage copy-frameworks` to 'New Run Script Phase'.
- Add `$(SRCROOT)/Carthage/Build/iOS/SwiftyPickerPopover.framework` to 'Input Files'.

### import

On your .swift file, import the module:

```swift
import SwiftyPickerPopover
```

### Popover types

SwiftyPickerPopover offers you the following popovers:

- **StringPickerPopover**: has an UIPickerView that allows user to choose a String type choice.
- **ColumnStringPickerPopover**: has an UIPickerView of multiple columns.
- **DatePickerPopover**: has an UIDatePicker that allows user to choose a Date type choice.
- **CountdownPickerPopover**: has an UIDatePicker specializing in countDownTimer style.

### APIs and examples

#### Common

All popovers have the following APIs.

* setPermittedArrowDirections()
* setArrowColor()
* setSize(width:,height:)
* setCornerRadius()
* setValueChange(action:)
* setOutsideTapDismissing(allowed:)
* setDimmedBackgroundView(enabled:)

* appear(originView:, baseViewWhenOriginViewHasNoSuperview:, baseViewController:, completion:)
* appear(barButtonItem:, baseViewWhenOriginViewHasNoSuperview:, baseViewController:, completion:)
* disappear()
* disappearAutomatically(after seconds:, completion:)
* reload()

#### StringPickerPopover

* init(title:, choices:)

* setFont()
* setFontSize()
* setFontColor()
* setImageNames()
* setImages()
* setSelectedRow()
* setRowHeight()
* setDisplayStringFor()
* setDoneButton(title:, font:, color:, action:)
* setCancelButton(title:, font:, color:, action:)
* setClearButton(title:, font:, color:, action:)

##### You can use StringPickerPopover like this:

```swift
StringPickerPopover(title: "StringPicker", choices: ["value 1", "value 2", "value 3"])
       .setSelectedRow(0)
       .setValueChange(action: { _, selectedDate in
               print("current date \(selectedDate)")
           
})
       .setDoneButton(action: { (popover, selectedRow, selectedString) in
    print("done row \(selectedRow) \(selectedString)")
})
       .setCancelButton(action: { (_, _, _) in print("cancel") }
)
        .appear(originView: button, baseViewController: self)
```

##### StringPickerPopover can have images.

<img src="README_resources/StringWithImage.jpeg" width="362">

After adding image files to your target's Assets.xcassets:

```swift
StringPickerPopover(title: "StringPicker", choices: ["value 1", "value2", ""])
       .setImageNames(["Icon1", nil, "Icon3"])
        .appear(originView: button, baseViewController: self)
```

##### It can separate the screen values from the raw values:

```swift
let displayStringFor: ((String?) -> String?)? = { string in
    if let s = string {
        switch (s) {
        case "value 1":
            return "😊"
        case "value 2":
            return "😏"
        case "value 3":
            return "😓"
        default:
            return s
        }
    }
    return nil
}

let p = StringPickerPopover(title: "StringPicker", choices: ["value 1", "value 2", "value 3"])
        .setDisplayStringFor(displayStringFor)
        .setDoneButton(action: {
            popover, selectedRow, selectedString in
            print("done row \(selectedRow) \(selectedString)")
        })
           .setCancelButton(action: { _, _, _ in
    print("cancel")
})

       p.appear(originView: sender, baseViewController: self)
       p.disappearAutomatically(after: 3.0, completion: { print("automatically hidden") })
```

##### To specify the size:

```swift
StringPickerPopover(title: "Narrow StringPicker", choices: ["value 1", "value 2", "value 3"])
           .setSize(width: 250.0)
           .appear(originView: sender, baseViewController: self)
```

The default width and height of popover are both 300.0. By using setSize(width:, height:), you can override it or them.
When you set nil to the parameter or don't specify it, the default will be used.

##### It appears from the collectionView's cell:

```swift
func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {

    let theCell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath)

           let p = StringPickerPopover(title: "CollectionView", choices: ["value 1", "value 2", "value 3"])
            .setSelectedRow(1)
            .setDoneButton(title: "👌", action: { (popover, selectedRow, selectedString) in print("done row \(selectedRow) \(selectedString)") })
                           .setCancelButton(title: "🗑", action: { (_, _, _) in print("cancel") })

           p.appear(originView: theCell, baseViewWhenOriginViewHasNoSuperview collectionView, baseViewController: self)

}
```

If originView has no superView, then then you need to set baseViewWhenOriginViewHasNoSuperview as above to specify
sourceView for the position for the arrow. If it has the superview, then SwiftyPickerPopover automatically use it for
the sourceView.

##### It appears from an UIBarButtonItem:

```swift
let item: UIBarButtonItem = sender
let originView = item.value(forKey: "view") as! UIView
p.appear(originView: originView, baseViewController: self)
```

#### ColumnStringPickerPopover

* init(title:, choices:, selectedRows:, columnPercents:)

* setFonts()
* setFontSizes()
* setFontColors()
* setSelectedRows()
* setDisplayStringFor()
* setDoneButton(title:, font:, color:, action:)
* setCancelButton(title:, font:, color:, action:)
* setClearButton(title:, font:, color:, action:)

##### ColumnStringPickerPopover can have multiple String values:

```swift
ColumnStringPickerPopover(title: "Columns Strings",
        choices: [["Breakfast", "Lunch", "Dinner"], ["Tacos", "Sushi", "Steak", "Waffles", "Burgers"]],
        selectedRows: [0, 0], columnPercents: [0.5, 0.5])
        .setDoneButton(action: { popover, selectedRows, selectedStrings in print("selected rows \(selectedRows) strings \(selectedStrings)") })
       .setCancelButton(action: { _, _, _ in print("cancel") })
        .setFontSize(14)
        .appear(originView: sender, baseViewController: self)
)
```

#### DatePickerPopover

* init(title:)

* setSelectedDate()
* setDateMode()
* setPreferredDatePickerStyle() - Available since iOS 13.4
* setMinimumDate()
* setMaximumDate()
* setMinuteInterval()
* setLocale()
* setDoneButton(title:, font:, color:, action:)
* setCancelButton(title:, font:, color:, action:)
* setClearButton(title:, font:, color:, action:)

##### DatePickerPopover can be used like this:

```swift
DatePickerPopover(title: "DatePicker")
        .setDateMode(.date)
        .setPreferredDatePickerStyle(.wheels)
        .setSelectedDate(Date())
        .setDoneButton(action: { popover, selectedDate in print("selectedDate \(selectedDate)") })
           .setCancelButton(action: { _, _ in print("cancel") })
        .appear(originView: sender, baseViewController: self)
```

##### The clear button rewinds the picker. And it disappers automatically after a specified seconds:

```swift
let p = DatePickerPopover(title: "Clearable DatePicker")
        .setPreferredDatePickerStyle(.wheels)
        .setDoneButton(action: { popover, selectedDate in print("selectedDate \(selectedDate)") })
           .setCancelButton(action: { _, _ in print("cancel") })
           .setClearButton(action: { popover, selectedDate in
    print("clear")
    //Rewind
    popover.setSelectedDate(Date()).reload()
})

p.appear(originView: sender, baseViewController: self)
p.disappearAutomatically(after: 3.0)
```

##### The time interval is 5 mins. The arrow is permitted only to .down direction.:

```swift
DatePickerPopover(title: "DatePicker .time 5minInt.")
        .setDateMode(.time)
        .setPreferredDatePickerStyle(.wheels)
        .setMinuteInterval(5)
        .setPermittedArrowDirections(.down)
        .setDoneButton(action: { popover, selectedDate in print("selectedDate \(selectedDate)") })
           .setCancelButton(action: { _, _ in print("cancel") })
        .appear(originView: sender, baseViewController: self)
)
```

#### CountdownPickerPopover

* init(title:)

* setSelectedTimeInterval
* setDoneButton(title:, font:, color:, action:)
* setCancelButton(title:, font:, color:, action:)
* setClearButton(title:, font:, color:, action:)

##### CountdownPickerPopover can be used like this:

```swift
 CountdownPickerPopover(title: "CountdownPicker")
        .setSelectedTimeInterval(TimeInterval())
        .setDoneButton(action: { popover, timeInterval in print("timeInterval \(timeInterval)") })
           .setCancelButton(action: { _, _ in print("cancel") })
           .setClearButton(action: { popover, timeInterval in
    print("Clear")
    popover.setSelectedTimeInterval(TimeInterval()).reload()
})
        .appear(originView: sender, baseViewController: self)
```

## Customize

### How do I customize a popover's storyboard?

When you prepare your customized Storyboard, it will be applied automatically.

1. Find the original file of the popover's Storyboard, which you want to change. For example, '
   CountdownPickerPopover.storyboard'.
2. Add it to your project on Xcode for including it into mainBundle. At this time, please check 'Copy items if needed'.
   Do not change the file name.
3. Next, change the module specified by default. Open the Storyboard file on Xcode and uncheck 'Inherit From Target' in
   InterfaceBuilder > Identity inspector > Custom Class. Specify 'SwiftyPickerPopover' for 'Module'.
4. Finally, customize your storyboard file.

## Contributors

- Ken Torimaru [GitHub](https://github.com/ktorimaru) for CountdownPickerPopover and ColumnStringPickerPopover.
- BalestraPatrick [GitHub](https://github.com/BalestraPatrick) for README.md typo.
- andersonlucasg3 [GitHub](https://github.com/andersonlucasg3) for adding possibility to override the storyboards with
  custom localizations in the app project.
- lswith [GitHub](https://github.com/lswith) for fixing circular reference issue of Cartfile.
- coybit [GitHub](https://github.com/coybit) for adding setImages() to StringPickerPopover.
- Mihael Isaev [GitHub](https://github.com/MihaelIsaev) for adding appear() from barButtonItem.
- iosMaher [GitHub](https://github.com/iosMaher) for idea of setFont() and setFontColor().
- gbuela [GitHub](https://github.com/gbuela) for setValueChange(action:) API for all popover types.
- ikbalyasar [GitHub](https://github.com/ikbalyasar) for demo code for showing selected value on StringPickerPopover.
- weakfl [GitHub](https://github.com/weakfl) for update to Swift 4.2
- Tobisaninfo[GitHub](https://github.com/Tobisaninfo( for cartage support with Xcode 10.2
- Alexwing[GitHub](https://github.com/Alexwings) for update to Swift 5.0

## Author

- Yuta Hoshino [Twitter](https://twitter.com/hsylife) [Facebook](https://www.facebook.com/yuta.hoshino)


CHANGELOG.md
6.6.6 Release notes (2019-10-03)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Nothing

### Bugfixes
* Update to Swift 5.0
* Support for Xcode 11


6.6.5 Release notes (2019-01-06)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Nothing

### Bugfixes
* Update to Swift 4.2

6.6.4 Release notes (2018-08-19)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* add: demo code for showing selected value on StringPickerPopover.

### Bugfixes
* Nothing

6.6.3 Release notes (2018-08-11)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Nothing

### Bugfixes
* fix: fontSize of StringPickerPopover do not work. fontSize has been accorded priority over font's pointSize.

6.6.2 Release notes (2018-06-21)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* add: TextFeildCell for demo

### Bugfixes
* fix: dimmedView does not disappear.

6.6.1 Release notes (2018-06-11)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Nothing

### Bugfixes
* fix: dismiss() is not called correctly.

6.6.0 Release notes (2018-06-03)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* add: setDimmedBackgroundView(enabled:)

### Bugfixes
* fix: arrow and naviBar are diffrent colors
* fix: circular references in appear() and disappearAutomatically()
* refactor: added VoidHandlerType


6.5.2 Release notes (2018-05-28)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Nothing

### Bugfixes

* fix: wierd view color of DatePickerPopover

6.5.1 Release notes (2018-05-27)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Nothing

### Bugfixes

* Refactor: polish code

6.5.0 Release notes (2018-05-27)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Add: setClearButton() to String/ColumnStringPickerPopover.

### Bugfixes

* Fix: setSelectedRow/Rows() don't work correctly.

6.4.0 Release notes (2018-05-24)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Add: setOutsideTapDismissing(allowed:) to popovers. It makes dismissing of popover by tapping outside of it avilable or not.

### Bugfixes

* Nothing

6.3.1 Release notes (2018-05-21)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Nothing

### Bugfixes

* Fix: popover navi bar color is different from parent navi bar color #83

6.3.0 Release notes (2018-05-20)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* add: font parameter to setDone/Cancle/ClearButton() for each popover #82

### Bugfixes

* Nothing

6.2.1 Release notes (2018-05-19)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Nothing

### Bugfixes

* Fix: crash when tapping done/cancel button with no items set String/ColumnStringPickerPopover #85
- Fix: wired arrow color when setting navigation bar color globally. #83


6.2.0 Release notes (2018-05-13)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Add: setValueChange(action:) for all popovers

### Bugfixes

* Nothing

6.1.0 Release notes (2018-02-09)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Add: setCornerRadius() for popover

### Bugfixes

* Nothing

6.0.1 Release notes (2018-02-02)
=============================================================

### API Breaking Changes

* Nothing

### Enhancements

* Added the new sample code about how to use StringPickerPopover with UITextField.

### Bugfixes

* Nothing

6.0.0 Release notes (2018-01-28)
=============================================================
### API Breaking Changes

* Change ColumnStringPickerPopover setFontSize() to setFontSizes().

### Enhancements

* Added setFont(), setFontColor() and setFontSize() to StringPickerPopover.
* Added setFonts(), setFontColors() and setFontSizes() to ColumnStringPickerPopover

### Bugfixes

*Fixed minor bugs.

5.3.0 Release notes (2017-08-04)
=============================================================

### API Breaking Changes

* None

### Enhancements

* Supports for adding UIImage to string directly

### Bugfixes

* None

5.2.3 Release notes (2017-08-03)
=============================================================

### API Breaking Changes

* None

### Enhancements

* None

### Bugfixes

* Remove cartfile to avoid circular reference when installing it via carthage

5.2.2 Release notes (2017-07-30)
=============================================================

### API Breaking Changes

* None

### Enhancements

* None

### Bugfixes

* Fix that cancelButton is set to nil on iOS 11

5.2.1 Release notes (2017-07-28)
=============================================================

### API Breaking Changes

* None

### Enhancements

* Add Podfile and Cartfile

### Bugfixes

* The demo target bundle identifier has been changed to "ge-nie-inc.SwiftyPickerPopoverDemo"

5.2.0 Release notes (2017-07-26)
=============================================================

### API Breaking Changes

* None

### Enhancements
* Add setLocale() to DatePickerPopover
* Improve comments and polish code

### Bugfixes

* None


5.1.2 Release notes (2017-07-23)

Demo/SampleViewController.swift
//
//  ViewController.swift
//  SwiftyPickerPopover
//
//  Created by Yuta Hoshino on 2016/09/14.
//  Copyright © 2016 Yuta Hoshino. All rights reserved.
//

import UIKit
import SwiftyPickerPopover

class SampleViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate {
    
    // for keeping selectedRow
    private var selectedRow: Int = 0
    
    @IBAction func tappedStringPickerButton(_ sender: UIButton) {
        /// Replace a string with the string to be display.
        let displayStringFor:((String?)->String?)? = { string in
            if let s = string {
                switch(s){
                case "value 1":
                    return "😊"
                case "value 2":
                    return "😏"
                case "value 3":
                    return "😓"
                default:
                    return s
                }
            }
            return nil
        }
        
        /// Create StringPickerPopover:
        let p = StringPickerPopover(title: "StringPicker", choices: ["value 1","value 2","value 3"])
            .setDisplayStringFor(displayStringFor)
            .setValueChange(action: { _, _, selectedString in
                print("current string: \(selectedString)")
            })
            .setFontSize(16)
            .setDoneButton(
                font: UIFont.boldSystemFont(ofSize: 16),
                color: UIColor.orange,
                action: { popover, selectedRow, selectedString in
                    print("done row \(selectedRow) \(selectedString)")
                    self.selectedRow = selectedRow
                    
            })
            .setCancelButton(action: {_, _, _ in
                print("cancel") })
        .setSelectedRow(selectedRow)
        p.appear(originView: sender, baseViewController: self)
        p.disappearAutomatically(after: 3.0, completion: { print("automatically hidden")} )
    }
    
    @IBAction func didTapStringPickerWithImageButton(_ sender: UIButton) {
        /// StringPickerPopover with image:
        let p = StringPickerPopover(title: "with image", choices: ["value 1","value 2",""])
            .setImageNames(["imageIcon",nil,"thumbUpIcon"])
            .setSize(width: 280)
            .setCornerRadius(0)
            .setValueChange(action: { _, _, selectedString in
                print("current string: \(selectedString)")
            })
            .setDoneButton(action: {
                popover, selectedRow, selectedString in
                print("done row \(selectedRow) \(selectedString)")
            })
            .setCancelButton(action: {_, _, _ in
                print("cancel") })
            .setOutsideTapDismissing(allowed: false)
            .setDimmedBackgroundView(enabled: true)
        p.appear(originView: sender, baseViewController: self)
    }
    
    @IBAction func didTapStringPickerClearableButton(_ sender: UIButton) {
        /// StringPickerPopover Clearable:
        let p = StringPickerPopover(title: "Clearable", choices: ["value 1","value 2","value3"])
            .setFont(UIFont.boldSystemFont(ofSize: 30))
            .setFontColor(.orange)
            .setFontSize(14)
            .setDoneButton(color: UIColor.red, action: {
                popover, selectedRow, selectedString in
                print("done row \(selectedRow) \(selectedString)")
            })
            .setCancelButton(action: {_, _, _ in
                print("cancel") })
            .setClearButton(title: "Reset", action: {(popover, row, value) in
                print("clear")
            })
        p.appear(originView: sender, baseViewController: self)
    }
    
    @IBAction func didTapStringPickerWithTextField(_ sender: UITextField) {
        StringPickerPopover(title: "TextField", choices: ["","Text 1", "Text 2", "Text 3"])
        .setValueChange(action: { _, _, selectedString in
            print("current string: \(selectedString)")
        })
        .setDoneButton(action: { popover, selectedRow, selectedString in
            sender.text = selectedString
        })
        .appear(originView: sender, baseViewController: self)        
    }
    
    @IBAction func tappendDatePickerButton(_ sender: UIButton) {
        /// DatePickerPopover appears:
        let picker = DatePickerPopover(title: "DatePicker")
            .setDateMode(.date)
            .setSelectedDate(Date())
            .setValueChange(action: { _, selectedDate in
                print("current date \(selectedDate)")
            })
            .setDoneButton(action: { popover, selectedDate in print("selectedDate \(selectedDate)")})
            .setCancelButton(action: { _, _ in print("cancel")})
        if #available(iOS 13.4, *) {
            picker.setPreferredDatePickerStyle(.wheels)
        }
        picker.appear(originView: sender, baseViewController: self)
    }
    
    @IBAction func tappendDatePickerCanClearButton(_ sender: UIButton) {
        /// DatePickerPopover appears:
        let p = DatePickerPopover(title: "Clearable DatePicker")
            .setLocale(identifier: "en_GB") //en_GB is dd-MM-YYYY. en_US is MM-dd-YYYY. They are both in English.
            .setValueChange(action: { _, selectedDate in
                print("current date \(selectedDate)")
            })
            .setDoneButton(action: { popover, selectedDate in print("selectedDate \(selectedDate)")} )
            .setCancelButton(action: { _, _ in print("cancel")})
            .setClearButton(action: { popover, selectedDate in
                print("clear")
                //Rewind
                popover.setSelectedDate(Date()).reload()
                //Instead, hide it.
//                popover.disappear()
            })
        if #available(iOS 13.4, *) {
            p.setPreferredDatePickerStyle(.wheels)
        }
        p.appear(originView: sender, baseViewController: self)
        p.disappearAutomatically(after: 3.0)
    }
    
    @IBAction func tappendDatePickerTime5MinIntButton(_ sender: UIButton) {
        // DatePickerPopover appears:
        let picker = DatePickerPopover(title: "DatePicker .time 5minInt.")
            .setDateMode(.time)
            .setMinuteInterval(5)
            .setPermittedArrowDirections(.down)
            .setValueChange(action: { _, selectedDate in
                print("current date \(selectedDate)")
            })
            .setDoneButton(action: { popover, selectedDate in print("selectedDate \(selectedDate)")} )
            .setCancelButton(action: { _, _ in print("cancel")})
        if #available(iOS 13.4, *) {
            picker.setPreferredDatePickerStyle(.wheels)
        }
        picker.appear(originView: sender, baseViewController: self)
    }

    @IBAction func countdownButton(_ sender: UIButton) {
        // CountdownPickerPopover appears:
        CountdownPickerPopover(title: "CountdownPicker")
            .setSelectedTimeInterval(TimeInterval())
            .setValueChange(action: { _, timeInterval in
                print("current interval \(timeInterval)")
            })
            .setDoneButton(action: { popover, timeInterval in print("timeInterval \(timeInterval)")} )
            .setCancelButton(action: { _, _ in print("cancel")})
            .setClearButton(action: { popover, timeInterval in print("Clear")
                popover.setSelectedTimeInterval(TimeInterval()).reload()
            })
            .appear(originView: sender, baseViewController: self)
    }
    
    @IBAction func columnsString(_ sender: UIButton) {
        //ColumnStringPickerPopover appears.
        ColumnStringPickerPopover(title: "Columns Strings",
                                  choices: [["Breakfast", "Lunch", "Dinner"], ["Tacos", "Sushi", "Steak", "Waffles", "Burgers"]],
                                  selectedRows: [0,0], columnPercents: [0.5, 0.5])
        .setDoneButton(action: { popover, selectedRows, selectedStrings in
            print("selected rows \(selectedRows) strings \(selectedStrings)")
        })
        .setCancelButton(action: { _,_,_ in print("cancel")})
        .setValueChange(action: { _, _, selectedStrings in
            print("current strings: \(selectedStrings)")
        })
        .setFonts([UIFont.boldSystemFont(ofSize: 14), nil])
        .setFontColors([nil, .red])
        .setFontSizes([20, nil]) // override
        .setSelectedRows([0,2])
        .appear(originView: sender, baseViewController: self)
    }

    //CollectionView
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell{
        
        let theCell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath)
        let label = theCell.contentView.viewWithTag(1) as! UILabel
        label.text = (indexPath as NSIndexPath).row.description
        
        return theCell
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        
        let theCell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath)
        
        //StringPickerPopover appears from the cell of collectionView.
        let p = StringPickerPopover(title: "Cell "+(indexPath as NSIndexPath).row.description, choices: ["value 1","value 2","value 3"])
        .setSelectedRow(1)
        .setValueChange(action: { _, _, selectedString in
            print("current string: \(selectedString)")
        })
        .setDoneButton(title:"👌", action: { (popover, selectedRow, selectedString) in
            print("done row \(selectedRow) \(selectedString)")
        })
        .setCancelButton(title:"👎", action: { _,_,_ in print("cancel")} )
        
        p.appear(originView: theCell, baseViewWhenOriginViewHasNoSuperview: collectionView, baseViewController: self)
        
    }
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 5
    }
}

